substrate/primitives/runtime/src/generic/checked_extrinsic.rs

```diff
@@ -18,81 +18,130 @@
 //! Generic implementation of an extrinsic that has passed the verification
 //! stage.
 
+use codec::Encode;
+use sp_weights::Weight;
+
 use crate::{
 	traits::{
-		self, DispatchInfoOf, Dispatchable, MaybeDisplay, Member, PostDispatchInfoOf,
-		SignedExtension, ValidateUnsigned,
+		self, transaction_extension::TransactionExtension, DispatchInfoOf, DispatchTransaction,
+		Dispatchable, MaybeDisplay, Member, PostDispatchInfoOf, ValidateUnsigned,
 	},
 	transaction_validity::{TransactionSource, TransactionValidity},
 };
 
+/// The kind of extrinsic this is, including any fields required of that kind. This is basically
+/// the full extrinsic except the `Call`.
+#[derive(PartialEq, Eq, Clone, sp_core::RuntimeDebug)]
+pub enum ExtrinsicFormat<AccountId, Extension> {
+	/// Extrinsic is bare; it must pass either the bare forms of `TransactionExtension` or
+	/// `ValidateUnsigned`, both deprecated, or alternatively a `ProvideInherent`.
+	Bare,
+	/// Extrinsic has a default `Origin` of `Signed(AccountId)` and must pass all
+	/// `TransactionExtension`s regular checks and includes all extension data.
+	Signed(AccountId, Extension),
+	/// Extrinsic has a default `Origin` of `None` and must pass all `TransactionExtension`s.
+	/// regular checks and includes all extension data.
+	General(Extension),
+}
+
+// TODO: Rename ValidateUnsigned to ValidateInherent
+// TODO: Consider changing ValidateInherent API to avoid need for duplicating validate
+//   code into pre_dispatch (rename that to `prepare`).
```

Comment:
I don't understand this. I thought `ValidateUnsigned` would just be removed.

What is the usage for those "inherents" ?

Right now there is provided inherents, with `ProvideInherent` trait implementation. they don't have any logic in `ValidateUnsigned`, their checks are in `check_inherent`.
Then there is some extrinsic like report equivocations or `submit_unsigned` for election. Those are very similar to inherent as their source is the node and they are invalid when coming from external source.
But we actually want them to resubmit IMO, contrary to provided inherents.

Also the solution I proposed offline doesn't work.
(The solution was to have those unsigned as "general" transaction where the node signs the call, and are only accepted if signed by block author. It doesn't work because it doesn't prevent replay. Like such extrinsic if put inside a block, the block is sent, people can see this extrinsic and now just spam the original node with its own extrinsic.)

I think there will be needed some work for those unsigned transactions from offchain worker. But we can do in a follow-up.

The solution will be one of:
* extend `ProvideInherent`,
* extend this `ValidateInherent` and modify the transaction pool to resubmit those inherents.
* Have `TransactionSource` in validate in general transaction.

Reply:
These comments were old and outdated. I removed most of them and left some relevant ones in place. In short, `ValidateUnsigned` will go away entirely and we will rely on extensions and the `General` transaction format to support these currently unsigned transactions, as discussed offline and outlined in extrinsic horizon.